### PR TITLE
[SERVICES-2242] Query for price candlestick data

### PR DESCRIPTION
**Pull request by mad2sm0key**
## Reasoning
- need data for plotting tradingview charts 

  
## Proposed Changes
- create hierarchical continuous aggregates of price data (pairs and tokens) 
- add a query that returns price candles for a specified interval and resolution

## How to test
-  when `series` is a token identifier :
```
query {
  priceCandles(
    series: "WEGLD-a28c59",
    metric: "priceUSD",
    start:"1711594835",
    end:"1711850435",
    resolution: MINUTE_1
  ) {
    time
    ohlc
  }
}
```
- when `series` is a pair address, metric can be `firstTokenPrice` or `secondTokenPrice`
```
query {
  priceCandles(
    series: "erd1qqqqqqqqqqqqqpgqus9r9gwtg24a9fvzv743hgydecpkxs8q0n4szz2az0",
    metric: "firstTokenPrice",
    start:"1711594835",
    end:"1711850435",
    resolution: HOUR_4
  ) {
    time
    ohlc
  }
}
```

### CODE REVIEW

**Review by claudiulataretu — APPROVED** [review]

File: src/modules/analytics/validators/price.candles.args.validator.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    [PriceCandlesResolutions.DAY_7]: MaxMinutesPerResolution.DAY_7,
+    [PriceCandlesResolutions.MONTH_1]: MaxMinutesPerResolution.MONTH_1,
+};
+
+@Injectable()
+export class PriceCandlesArgsValidationPipe implements PipeTransform {
+    async transform(value: PriceCandlesQueryArgs, metadata: ArgumentMetadata) {
+        const { start, end, resolution } = value;
+
+        if (!this.isValidTimestamp(start)) {
```
Comment (bogdan-rosianu) [inline src/modules/analytics/validators/price.candles.args.validator.ts:38]: isn't this validated via `@IsValidUnixTime()` ? Or are there cases when you can bypass the check so you needed to make it here as well?

File: src/modules/analytics/validators/price.candles.args.validator.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            throw new UserInputError('Invalid timestamp format for start.');
+        }
+
+        const startDate = moment.unix(parseInt(start));
+        const maxDuration = resolutionToMaxDuration[resolution];
+
+        if (!end) {
+            const currentTime = moment.utc();
```
Comment (bogdan-rosianu) [inline src/modules/analytics/validators/price.candles.args.validator.ts:46]: could have made a new function that computes the end-time

File: src/modules/analytics/validators/price.candles.args.validator.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+export class PriceCandlesArgsValidationPipe implements PipeTransform {
+    async transform(value: PriceCandlesQueryArgs, metadata: ArgumentMetadata) {
+        const { start, end, resolution } = value;
+
+        if (!this.isValidTimestamp(start)) {
+            throw new UserInputError('Invalid timestamp format for start.');
+        }
+
+        const startDate = moment.unix(parseInt(start));
+        const maxDuration = resolutionToMaxDuration[resolution];
```
Comment (bogdan-rosianu) [inline src/modules/analytics/validators/price.candles.args.validator.ts:43]: might check if `resolution` exists in the map

File: src/modules/analytics/validators/price.candles.args.validator.ts (hunk 1 above)
Comment (mad2sm0key) [reply src/modules/analytics/validators/price.candles.args.validator.ts:38]: it was redundant indeed. removed the decorator

File: src/modules/analytics/validators/price.candles.args.validator.ts (hunk 2 above)
Comment (mad2sm0key) [reply src/modules/analytics/validators/price.candles.args.validator.ts]: done

File: src/modules/analytics/validators/price.candles.args.validator.ts (hunk 3 above)
Comment (mad2sm0key) [reply src/modules/analytics/validators/price.candles.args.validator.ts:43]: there is no need for it. Graphql performs enum validation already

**Review by bogdan-rosianu — APPROVED** [review]

**Review by claudiulataretu — APPROVED** [review]

**Pull request merged**